### ISSUE TITLE
About the softmax operation

### ISSUE THREAD

**Issue by conditionWang** [hongjoon0805/SS-IL-Official#2]
Hi, thanks for releasing the code. I notice that there are two Separated Softmax operations for the seen tasks and the current task in your ICCV 2021 version. However, in your code, especially for the ssil.py line 66 to 94, it seems that there is not any softmax operation in the computation of loss_CE. I am implementing your code and want to cite it as a baseline in my research, but the performance of your ssil method is even worse than that of iCaRL. I am not sure whether there is any problem.

**Comment by hongjoon0805** [comment]
Hi,
Thank you for giving interest to our algorithm.

In self.loss = torch.nn.CrossEntropyLoss, it internally computes the softmax of given logit, and makes the loss function based on the target and the softmax.

Because our algorithm uses separated softmax, we sliced the output logit like curr = output[:batch_size,mid:end], and we give it to self.loss.

Therefore, it seems we did not use any softmax operation externally, the self.loss internally computes the softmax.

If there are more questions, please feel free to contact us.

Thank you,

Hongjoon

**Comment by conditionWang** [comment]
Hi Hongjoon,

I notice you mention the 'logit', regularly, the logit is the computation result after feeding the output to sigmoid. In your implementation, the output won't be conducted sigmoid operation before CrossEntropyLoss computation, is that right? Btw, I still get a worse performance of your ssil than iCaRL, even with the same hyper-parameter setting with your experiment. Is there any tip that I can improve the performance? I have checked several times for my implementation.

Best,
Lixu

**Comment by hongjoon0805** [comment]
In our implementation, the logit is the value before the softmax or sigmoid, and its value is passed to the CrossEntropyLoss. In PyTorch, CrossEntropyLoss doesn't need the softmax operated value because it internally do that. 

For the performance of your implementation, did you run our official code or your implementation?
If you run your implementation, I'll give some tips for implementing SS-IL

1. Please be noticed that our implementation uses torch.nn.CrossEntropyLoss(reduction='sum'). We do not use torch.nn.CrossEntropyLoss(reduction='mean')
2. You should carefully slice the output value based on the number of current / previous classes
3. After slicing the output, its value is passed to the self.loss, and we manually averaged the loss like loss_CE = (loss_CE_curr + loss_CE_prev) / (batch_size + replay_size)

In our paper, we showed that SS-IL w/o TKD also performs well, and the average accuracy also outperforms iCaRL.

If there are any questions, please contact us.

Thank you,
Hongjoon

**Comment by conditionWang** [comment]
Thanks so much for your reply, I have sent you an email. I think I need further help in implementing your method.

**Comment by qsunyuan** [comment]
https://github.com/hongjoon0805/SS-IL-Official/blob/8c323811f99f6da8e8d794b663c43bfe63864a3f/trainer/icarl.py#L42

In icarl, already separated Softmax operations?

**Comment by qsunyuan** [comment]
line 42 - line 52. I'm not sure.

**Comment by hongjoon0805** [comment]
Hi,

line 42 is for task-wise KD (TKD). Therefore, we use separated softmax for KD.

Thanks,
Hongjoon

**Comment by qsunyuan** [comment]
thx for your quick replay.
I found the TKD is used in iCaRL in your codes. In my view, but this is not the same as the original iCaRL.
I am not sure how the current community implements this iCaRL baseline.

<img width="642" alt="1638350335(1)" src="https://user-images.githubusercontent.com/71776283/144206674-472c3209-e84d-4572-88ef-48334dd3dbe2.png">

**Comment by hongjoon0805** [comment]
The original implementation is using BCE loss for KD. But, we empirically find that using softmax instead of sigmoid function outperforms the original, so we just used the softmax.

**Comment by qsunyuan** [comment]
Thanks so much for your help.
